### PR TITLE
RHOS dedicated g/w node support

### LINKED ISSUE #167: Enhance cloud prepare to support dedicated g/w node in RHOS
**What would you like to be added**:
Enhance cloud prepare to support dedicated g/w node in RHOS
**Why is this needed**:
This needed to support Submariner in RHOS

### PR DESCRIPTION
**Pull request by aswinsuryan**
Signed-off-by: Aswin Suryanarayanan <aswinsuryan@gmail.com>

### CODE REVIEW

File: pkg/k8s/k8siface.go [hunk 1]
```diff
@@ -52,7 +54,7 @@ func NewInterface(clientSet kubernetes.Interface) Interface {
 func (k *k8sIface) ListNodesWithLabel(labelSelector string) (*v1.NodeList, error) {
 	nodes, err := k.clientSet.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to list the nodes in the cluster")
+		return nil, fmt.Errorf("unable to list the nodes in the cluster, err: %w", err)
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: Why was this changed?

File: pkg/k8s/k8siface.go [hunk 2]
```diff
@@ -63,31 +65,32 @@ func (k *k8sIface) ListGatewayNodes() (*v1.NodeList, error) {
 
 	nodes, err := k.clientSet.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{LabelSelector: labelSelector})
 	if err != nil {
-		return nil, errors.Wrap(err, "unable to list the Gateway nodes in the cluster")
+		return nil, fmt.Errorf("unable to list the Gateway nodes in the cluster, err: %w", err)
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: ...and this?

File: pkg/k8s/k8siface.go [hunk 3]
```diff
@@ -63,31 +65,32 @@ func (k *k8sIface) ListGatewayNodes() (*v1.NodeList, error) {
...
 	client := &resource.InterfaceFuncs{
 		GetFunc: func(ctx context.Context, name string, options metav1.GetOptions) (runtime.Object, error) {
-			return k.clientSet.CoreV1().Nodes().Get(ctx, name, options)
+			node, err := k.clientSet.CoreV1().Nodes().Get(ctx, name, options)
+			return node, errors.WithMessagef(err, "error getting node %q", name)
 		},
 		UpdateFunc: func(ctx context.Context, obj runtime.Object, options metav1.UpdateOptions) (runtime.Object, error) {
-			return k.clientSet.CoreV1().Nodes().Update(ctx, obj.(*v1.Node), options)
+			node, err := k.clientSet.CoreV1().Nodes().Update(ctx, obj.(*v1.Node), options)
+			return node, errors.WithMessagef(err, "error updating node %q ", obj.(*v1.Node).Name)
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: As per the prior comment, I don't think we need to wrap here as these functions are pass-throughs.

File: pkg/k8s/k8siface.go [hunk 4]
```diff
@@ -63,31 +65,32 @@ func (k *k8sIface) ListGatewayNodes() (*v1.NodeList, error) {
...
 		},
 	}
 
-	return errors.Wrap(util.Update(context.TODO(), client, &v1.Node{
+	return errors.WithMessagef(util.Update(context.TODO(), client, &v1.Node{
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: Why was this changed from `Wrap` to `WithMessage`? Same in other places?

File: pkg/rhos/ocpgwdeployer.go [hunk 5]
```diff
@@ -18,28 +18,126 @@ limitations under the License.
...
+
+	unstructDecoder := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
+	machineSet := &unstructured.Unstructured{}
+	_, _, err = unstructDecoder.Decode(gatewayYAML, nil, machineSet)
+
+	if err != nil {
+		return nil, errors.WithMessage(err, "error decoding message gateway yaml")
+	}
+
+	return machineSet, nil
```
Comment (tpantelis) [inline pkg/rhos/ocpgwdeployer.go]:
```suggestion
	return machineSet, errors.WithMessage(err, "error decoding message gateway yaml")
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/k8s/k8siface.go [hunk 6]
```diff
@@ -70,13 +72,14 @@ func (k *k8sIface) ListGatewayNodes() (*v1.NodeList, error) {
...
 	client := &resource.InterfaceFuncs{
 		GetFunc: func(ctx context.Context, name string, options metav1.GetOptions) (runtime.Object, error) {
-			return k.clientSet.CoreV1().Nodes().Get(ctx, name, options)
+			node, err := k.clientSet.CoreV1().Nodes().Get(ctx, name, options)
+			return node, errors.Wrap(err, fmt.Sprintf("error getting node %q", name))
 		},
 		UpdateFunc: func(ctx context.Context, obj runtime.Object, options metav1.UpdateOptions) (runtime.Object, error) {
-			return k.clientSet.CoreV1().Nodes().Update(ctx, obj.(*v1.Node), options)
+			node, err := k.clientSet.CoreV1().Nodes().Update(ctx, obj.(*v1.Node), options)
+			return node, errors.Wrap(err, fmt.Sprintf("error updating node %q ", obj.(*v1.Node).Name))
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: As mentioned before, these are pass-throughs so we don't need to wrap - let the caller do it.

File: pkg/k8s/k8siface.go [hunk 7]
```diff
@@ -105,19 +108,22 @@ func (k *k8sIface) AddGWLabelOnNode(nodeName string) error {
...
-		err = k.updateLabel(node.Name, func(existing *v1.Node) {
-			delete(existing.Labels, SubmarinerGatewayLabel)
-		})
-
+	gwNodes := gwNodeList.Items
+	for i := range gwNodes {
+		err = k.RemoveGWLabelFromWorkerNode(&gwNodes[i])
 		if err != nil {
-			return err
+			return errors.Wrap(err, fmt.Sprintf("error removing the label from the gateway node %q", gwNodes[i].Name))
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go:117]:
```suggestion
			return errors.Wrapf(err, "error removing the label from the gateway node %q", gwNodes[i].Name)
```

File: pkg/ocp/machinesets.go [hunk 8]
```diff
@@ -89,17 +88,24 @@ func (msd *k8sMachineSetDeployer) GetWorkerNodeImage(machineSet *unstructured.Un
...
+				image, _, _ := unstructured.NestedString(disk, "image")
+				if image != "" {
+					return image, nil
+				}
+			}
 		}
 	}
 
-	return "", fmt.Errorf("could not find the image of one of the worker nodes on GCP infra %q", infraID)
+	return "", errors.Wrap(err, fmt.Sprintf("could not retrieve the image of one of the worker nodes on gcp infra %q", infraID))
```
Comment (tpantelis) [inline pkg/ocp/machinesets.go]: There's no error in scope that would be non-nil here - `fmt.Errorf` is appropriate here..

File: pkg/rhos/ocpgwdeployer.go [hunk 9]
```diff
@@ -65,55 +159,73 @@ func (d *ocpGatewayDeployer) Deploy(input api.GatewayDeployInput, reporter api.R
...
+				alreadyTagged := nodes[i].GetLabels()[submarinerGatewayNodeTag]
+				if alreadyTagged == "true" {
+					continue
+				}
+
+				reporter.Started(fmt.Sprintf("Configuring worker node %q as Submariner gateway node", nodes[i].Name))
+
+				err := d.K8sClient.AddGWLabelOnNode(nodes[i].Name)
+				if err != nil {
+					return errors.Wrap(err, fmt.Sprintf("failed to label the node %q as Submariner gateway node", nodes[i].Name))
```
Comment (tpantelis) [inline pkg/rhos/ocpgwdeployer.go]:
```suggestion
					return errors.Wrapf(err, "failed to label the node %q as Submariner gateway node", nodes[i].Name)
```

File: pkg/rhos/ocpgwdeployer.go [hunk 10]
```diff
@@ -125,28 +237,51 @@ func (d *ocpGatewayDeployer) Cleanup(reporter api.Reporter) error {
 
 	computeClient, err := openstack.NewComputeV2(d.Client, gophercloud.EndpointOpts{Region: d.Region})
 	if err != nil {
-		return errors.WithMessagef(err, "error creating the compute client for the region: %q", d.Region)
+		return errors.Wrap(err, fmt.Sprintf("error creating the compute client for the region: %q", d.Region))
```
Comment (tpantelis) [inline pkg/rhos/ocpgwdeployer.go]:
```suggestion
		return errors.Wrapf(err, "error creating the compute client for the region: %q", d.Region)
```

File: pkg/rhos/ocpgwdeployer.go [hunk 11]
```diff
@@ -125,28 +237,51 @@ func (d *ocpGatewayDeployer) Cleanup(reporter api.Reporter) error {
...
-		return errors.WithMessage(err, "failed to remove labels from worker node")
+		// If the instance name matches with d.InfraID + "-submariner-gw-", it implies that
+		// the gateway node was deployed using the OCPMachineSet API otherwise it's an existing worker node.
+		prefix := d.InfraID + "-submariner-gw-"
+		if strings.HasPrefix(gwNodes[i].Name, prefix) {
+			reporter.Started(fmt.Sprintf("Deleting the gateway instance %q", gwNodes[i].Name))
+
+			err = d.deleteGateway(strconv.Itoa(i))
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("error deleting the Submariner gateway security group rules from node: %q",
```
Comment (tpantelis) [inline pkg/rhos/ocpgwdeployer.go]:
```suggestion
				return errors.Wrapf(err, "error deleting the Submariner gateway security group rules from node: %q",
```

File: pkg/rhos/ocpgwdeployer.go [hunk 12]
```diff
@@ -125,28 +237,51 @@ func (d *ocpGatewayDeployer) Cleanup(reporter api.Reporter) error {
...
+					gwNodes[i].Name))
+			}
+
+			reporter.Succeeded("Successfully deleted the instance")
+		} else {
+			err = d.removeGWFirewallRules(groupName, gwNodes[i].Name, computeClient)
+			if err != nil {
+				return errors.Wrap(err, fmt.Sprintf("error deleting the Submariner gateway security group rules from node: %q",
```
Comment (tpantelis) [inline pkg/rhos/ocpgwdeployer.go]:
```suggestion
				return errors.Wrapf(err, "error deleting the Submariner gateway security group rules from node: %q",
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/k8s/k8siface.go (hunk 6 above)
Comment (aswinsuryan) [reply pkg/k8s/k8siface.go]: Linter was throwing an error if not wrapped

pkg/k8s/k8siface.go:81:11: error returned from interface method should be wrapped: sig: func (k8s.io/client-go/kubernetes/typed/core/v1.NodeInterface).Update(ctx context.Context, node *k8s.io/api/core/v1.Node, opts k8s.io/apimachinery/pkg/apis/meta/v1.UpdateOptions) (*k8s.io/api/core/v1.Node, error) (wrapcheck)
			return k.clientSet.CoreV1().Nodes().Update(ctx, obj.(*v1.Node), options)

Comment (tpantelis) [reply pkg/k8s/k8siface.go]: Right - that's why I previously had "_// nolint:wrapcheck // Let the caller wrap these errors._" on L73.

Comment (aswinsuryan) [reply pkg/k8s/k8siface.go]: oh ok i got it now, added the comment .

File: pkg/k8s/k8siface.go [hunk 13]
```diff
@@ -70,12 +72,13 @@ func (k *k8sIface) ListGatewayNodes() (*v1.NodeList, error) {
...
 
 func (k *k8sIface) updateLabel(nodeName string, mutate func(existing *v1.Node)) error {
-	// nolint:wrapcheck // Let the caller wrap these errors.
 	client := &resource.InterfaceFuncs{
 		GetFunc: func(ctx context.Context, name string, options metav1.GetOptions) (runtime.Object, error) {
+			// nolint:wrapcheck // Let the caller wrap these errors.
 			return k.clientSet.CoreV1().Nodes().Get(ctx, name, options)
 		},
 		UpdateFunc: func(ctx context.Context, obj runtime.Object, options metav1.UpdateOptions) (runtime.Object, error) {
+			// nolint:wrapcheck // Let the caller wrap these errors.
```
Comment (tpantelis) [inline pkg/k8s/k8siface.go]: This works but I meant to put the comment back before L75 where it was previously.

Comment (aswinsuryan) [reply pkg/k8s/k8siface.go]: I corrected it.

**Review by tpantelis — APPROVED** [review]

File: pkg/gcp/ocpgwdeployer.go [hunk 14]
```diff
@@ -251,7 +251,9 @@ func (d *ocpGatewayDeployer) deployGateway(zone string) error {
 	}
 
 	if d.image == "" {
-		d.image, err = d.msDeployer.GetWorkerNodeImage(machineSet, d.InfraID)
+		workerNodeList := []string{d.InfraID + "-worker-b", d.InfraID + "-worker-c", d.InfraID + "-worker-d"}
```
Comment (sridhargaddam) [inline pkg/gcp/ocpgwdeployer.go:255]: Can we safely assume that nodes will be labelled in this manner (i.e., `-worker-x`)?
Also, here we are hard-coding the list to just four nodes. Is there a better way to simply query the worker nodes based on the labels and pick one of them?

File: pkg/ocp/machinesets.go [hunk 15]
```diff
@@ -69,15 +69,14 @@ func (msd *k8sMachineSetDeployer) clientFor(obj runtime.Object) (dynamic.Resourc
...
-func (msd *k8sMachineSetDeployer) GetWorkerNodeImage(machineSet *unstructured.Unstructured, infraID string) (string, error) {
+func (msd *k8sMachineSetDeployer) GetWorkerNodeImage(workerNodeList []string, machineSet *unstructured.Unstructured,
+	infraID string) (string, error) {
 	machineSetClient, err := msd.clientFor(machineSet)
 	if err != nil {
 		return "", err
 	}
 
 	// TODO: After implementing a ListAll method in admiral, modify this code accordingly.
-	workerNodeList := []string{infraID + "-worker-b", infraID + "-worker-c", infraID + "-worker-d"}
```
Comment (sridhargaddam) [inline pkg/ocp/machinesets.go:79]: Can we remove the above TODO comment as it does not seem to be relevant any more?

File: pkg/ocp/machinesets.go [hunk 16]
```diff
@@ -89,17 +88,24 @@ func (msd *k8sMachineSetDeployer) GetWorkerNodeImage(machineSet *unstructured.Un
...
+				image, _, _ := unstructured.NestedString(disk, "image")
+				if image != "" {
+					return image, nil
+				}
+			}
 		}
 	}
 
-	return "", fmt.Errorf("could not find the image of one of the worker nodes on GCP infra %q", infraID)
+	return "", fmt.Errorf("could not retrieve the image of one of the worker nodes on rhos infra %q", infraID)
```
Comment (sridhargaddam) [inline pkg/ocp/machinesets.go]:
```suggestion
	return "", fmt.Errorf("could not retrieve the image of one of the worker nodes from the infra %q", infraID)
```

As this is used even in GCP, lets try to use some generic message.

**Review by sridhargaddam — CHANGES_REQUESTED** [review]

File: pkg/ocp/machinesets.go (hunk 16 above)
Comment (aswinsuryan) [reply pkg/ocp/machinesets.go]: Done.

File: pkg/ocp/machinesets.go (hunk 15 above)
Comment (aswinsuryan) [reply pkg/ocp/machinesets.go:79]: Moved the TODO to the respective callers.

File: pkg/gcp/ocpgwdeployer.go (hunk 14 above)
Comment (aswinsuryan) [reply pkg/gcp/ocpgwdeployer.go:255]: I was not sure why it was done this way , but as per the TODO it is till admiral adds an API.

Comment (tpantelis) [reply pkg/gcp/ocpgwdeployer.go:255]: I'm not sure I follow. Why what was done? Lines 254-255 are new, added by this PR. Can you be more specific re: a `ListAll` method in admiral, ie specifically where would it be added and what would it do?

Comment (aswinsuryan) [reply pkg/gcp/ocpgwdeployer.go:255]: These are moved from the pkg/ocp/machinesets.go to this file and now the workerNodeList is passed as argument  to msDeployer.GetWorkerNodeImage so that the method can be reused. This may be referring to a list all function to see all the worker nodes in the infra.

Comment (tpantelis) [reply pkg/gcp/ocpgwdeployer.go:255]: OK I still don't see where such a `ListAll` function would go or why it would be needed in admiral. Not trying to nit but unless we understand what the TODO means, I'd rather not propagate it.

Comment (aswinsuryan) [reply pkg/gcp/ocpgwdeployer.go:255]: I removed the TODO.

Comment (sridhargaddam) [reply pkg/gcp/ocpgwdeployer.go:255]: Basically, we are using the admiral dynamic Interface API to Get the machineSet details associated with the worker-node. Currently, the admiral API does not have the listAll method (https://github.com/submariner-io/admiral/blob/69a40598bdd63f570d0e695b2fd1a3ebdeb8f0f0/pkg/resource/interface.go#L28), hence we are using this approach to get the machineSet details of the associated worker nodes. If we have a listAll method in admiral, we can remove this hard-coding of the workerNodeList.

@aswinsuryan thanks for updating the PR. It looks good. Can you please add a TODO with the above notes so that once Admiral is enhanced we can update this code accordingly?

Comment (tpantelis) [reply pkg/gcp/ocpgwdeployer.go:255]: I'm confused. I don't see the admiral `Resource` interface being used anywhere in this code path. This code came from `k8sMachineSetDeployer.GetWorkerNodeImage` which uses a `dynamic.Interface` obtained from

    `machineSetClient, err := msd.clientFor(machineSet)`

 so the standard K8s List method is available.

Comment (sridhargaddam) [reply pkg/gcp/ocpgwdeployer.go:255]: Yes @tpantelis. I am not able to recollect the exact reason why that TODO was added. But it looks like we can use machineSetClient.List() method to get the list of worker nodes and remove the hardcoded list.

**Review by sridhargaddam — APPROVED** [review]

File: pkg/gcp/ocpgwdeployer.go (hunk 14 above)
Comment (aswinsuryan) [reply pkg/gcp/ocpgwdeployer.go:255]: I have added a TODO and raised this issue to track https://github.com/submariner-io/cloud-prepare/issues/201
